listenbrainz/webserver/static/js/src/recommendations/Recommendations.tsx

```diff
@@ -54,6 +56,7 @@ export default class Recommendations extends React.Component<
             props.recommendations.length / this.expectedRecommendationsPerPage
           )
         : 0,
+      recommendationFeedbackMap: {},
```

Comment:
umm...we basically need to do a lookup for an mbid to find the corresponding rating. I am not sure how can we store recording_mbid and corresponding rating. list of lists, something like [['mbid A', 'rating A']]? but that's too nested ig.

Reply:
or maybe have two arrays `recommendationFeedbackMbids` and `recommendationFeedbackRating` 
`recommendationFeedbackMbids`  = [mbid 1, mbid 2]
`recommendationFeedbackRating`  = [rating1, rating 2]
 where
`recommendationFeedbackMbids[0]`  corresponds to `recommendationFeedbackRating[0]`
and so on...

but this can be messy to maintain and error-prone ig

Reply:
You're right, I wrote too fast yesterday, forget what I said about arrays, it would indeed look awful :)

There are however a couple of things to keep in mind with nested properties in the state, but in short we want to make sure that 1. we make a copy of and modify the map (so that react detects a change)  and 2. We make sure we're not overwriting parts of the state inadvertantly (the calls to setstate are asynchronous).
The TL;DR is this syntax:
`this.setState(currentState => { recommendationFeedbackMap: { ...currentState. recommendationFeedbackMap, 'some-bbid': 0} });`
We give setState a callback function that is called with the previous state (this ensures we're not overwriting other things), make a  copy of the feedbackmap (the object spread syntax "…"), and modify the one element we want.

Some reading material:
https://medium.com/@sydbails/how-i-created-nested-properties-in-my-react-js-state-and-why-you-probably-shouldnt-7d9453340e04
https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data

Reply:
https://youtu.be/k5HR9gJh1rw 
Have taken help from this too. 

So what I understand is, since js/react is asynchronous and also react encourages to not mutate states, we want to copy previous state, mutate the copy and then set it as the current state. 

Since the lang is asynchronous, it might happen that when we mutate the previous state to assign it to current state, the call reverts or something happens and we want to render the previous state but it would lead to wrong results since we have already mutated the previous state. That is why it is encouraged to not mutate the state as such but make a copy. The way I imagine is by pointers, mutating the states without copy is like every state in the hierarchy is pointing to same location in the memory though what we ideally want is to have a separate location (copy of data) for every state. 


I hope I make sense (though I feel I have used the words "previous", " Current", "state" In a confusing manner)

Reply:
It would have been nice if typescript or the compiler points out that "hey, you are mutating states, be careful", something like that , will help noobs at least.

Reply:
>  it might happen that when we mutate the previous state to assign it to current state, the call reverts or something happens

Yes, that's one of the potential issues, nicely summarized :)
In our case, it might be unlikely that a user would click feedback icons so fast that we would run into that issue (click on 'love', quickly click on 'hate', if the async 'hate' setState finishes before the 'love' one finished, the 'love' feedback would be the last applied; confusing for the user…), but let's not code something that allows that.

The second potential issue is that is a bit harder to detect is that when React compares the state  `{foo: {bar:0, baz:1})` and `{foo: {bar:0, baz:2}}` it won't see any difference; for React, that's the same object with the same structure, so don't update anything; it won't make a deep comparison of each nested value because that can be really slow.
That's why we create a *new* `foo` object (using for example `foo2 = {…foo}` or `foo2 = Object.assign({}, foo)`) to ensure React knows something has changed, which will trigger a re-render.

Reply:
Right, so should we avoid stuff like this too? 
https://github.com/metabrainz/listenbrainz-server/blob/rec-component-feedback/listenbrainz/webserver/static/js/src/recommendations/Recommendations.tsx#L117

I have used it at many places to get the recs for the current page, and now I think it isn't the right way, maybe!

Reply:
Ahaaa, very interesting question!
In this case, you should be just fine. The reason is that array.slice returns a new array, so no chance for Readct to get confused. 
From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice :
>The slice() method returns a shallow copy of a portion of an array into a new array object

The other thing is that we're not dealing with a nested object here but instead an array. That should be absolutely fine.
It's if you had in your state an javascript object with nested properties (`setState({loading: true, recommendations: {nestedProperty1: true, nestedProperty2: false}})`) that you would have the same problem.